js/src/common/components/Button.tsx

```diff
@@ -0,0 +1,124 @@
+import Component, { ComponentAttrs } from '../Component';
+import fireDebugWarning from '../helpers/fireDebugWarning';
+import icon from '../helpers/icon';
+import classList from '../utils/classList';
+import extractText from '../utils/extractText';
+import LoadingIndicator from './LoadingIndicator';
+
+export interface ButtonAttrs extends ComponentAttrs {
+  /**
+   * Class(es) of an optional icon to be rendered within the button.
+   *
+   * If provided, the button will gain a `has-icon` class.
+   */
+  icon?: string;
+  /**
+   * Disables button from user input.
+   *
+   * Default: `false`
+   */
+  disabled?: boolean;
+  /**
+   * Show a loading spinner within the button.
+   *
+   * If `true`, also disables the button.
+   *
+   * Default: `false`
+   */
+  loading?: boolean;
+  /**
+   * **DEPRECATED:** Please use the `aria-label` attribute instead. For tooltips, use
+   * the `<Tooltip>` component.
+   *
+   * Accessible text for the button. This should always be present if the button only
+   * contains an icon.
+   *
+   * The textual content of this attribute is passed to the DOM element as `aria-label`.
+   *
+   * @deprecated
+   */
+  title?: string | Mithril.ChildArray;
+  /**
+   * Accessible text for the button. This should always be present if the button only
+   * contains an icon.
+   *
+   * The textual content of this attribute is passed to the DOM element as `aria-label`.
+   */
+  'aria-label'?: string | Mithril.ChildArray;
+  /**
+   * Button type.
+   *
+   * Default: `"button"`
+   *
+   * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type
+   */
+  type?: string;
+}
+
+/**
+ * The `Button` component defines an element which, when clicked, performs an
+ * action.
+ *
+ * Other attrs will be assigned as attributes on the `<button>` element.
+ *
+ * Note that a Button has no default class names. This is because a Button can
+ * be used to represent any generic clickable control, like a menu item. Common
+ * styles can be applied by providing `className="Button"` to the Button component.
+ */
+export default class Button extends Component<ButtonAttrs> {
+  view(vnode: Mithril.Vnode<ButtonAttrs, never>) {
+    let { type, title, 'aria-label': ariaLabel, icon: iconName, disabled, loading, className, class: _class, ...attrs } = this.attrs;
+
+    // If no `type` attr provided, set to "button"
+    type ||= 'button';
+
+    // Use `title` attribute as `aria-label` if none provided
+    ariaLabel ||= title;
+
+    // If given a translation object, extract the text.
+    if (typeof ariaLabel === 'object') {
+      ariaLabel = extractText(ariaLabel);
+    }
+
+    if (disabled || loading) {
+      delete attrs.onclick;
+    }
+
+    className = classList(_class, className, {
+      hasIcon: iconName,
+      disabled: disabled || loading,
+      loading: loading,
+    });
+
+    if (!ariaLabel && !extractText(vnode.children) && !this.element?.getAttribute?.('aria-label')) {
```

Comment:
Shouldn't the `aria-label` attribute of element be the same as `ariaLabel`? Why do you need to check both?

Reply:
ariaLabel is provided directly to the Button component, but we also need to see if `aria-label` is present on the DOM element as that's how the Tooltip component provides accessible labels.

Reply:
shouldn't it be present because aria label is passed to the <button> attributes?

Reply:
It's assigned to the actual DOM node's attributes, but the relationship between the DOM and VDOM in Mithril is one-way. If we update the actual DOM, the component's attributes don't change, as far as I'm aware.

https://github.com/flarum/core/blob/ce824b0ccff6a6570e2254e207e0d17fbbdd1b1c/js/src/common/components/Tooltip.tsx#L241